// bx-builder-angular/tools/grunt/config/concat.js
const _ = require('lodash');

module.exports = function(grunt, options){

    const config = grunt.config.get('config');

    const concat = {
        dev: {
            options: {
                sourceMap: false
            },
            files: {
            }
        },
        prod: {
            options: {
                sourceMap: false
            },
            files: {
            }
        }
    };

    // https://github.com/gruntjs/grunt-contrib-concat

    // concat files 을 초기화 한다.
    (function initConcatFiles(){

        if (!config.build['js']) return;

        const js = config.build['js'];

        // 목적지 경로
        const defaultPath =  '<%=config.tools.path.build%>';
        const subPath = '/common/';
        const distPath = defaultPath + subPath;
        const ext = 'js';

        if (js['external']){

            var external = js['external'];

            for (var loadType in external){

                if (!external[loadType]) continue;

                var src = [];

                for (var name in external[loadType]){

                    if (!external[loadType][name]['absolutePath']) continue;

                    src.push(external[loadType][name]['absolutePath']);
                }

                // external file 병합
                concat.dev.files[distPath + 'external.' + loadType + '.' + ext] = src;
                concat.prod.files[distPath + 'external.' + loadType + '.' + ext] = src;

            }
        }
    })();

    return concat;
};
