// bx-builder-angular/tools/grunt/config/htmlmin.js
/**
 * Created by sgjeon on 16. 4. 8..
 */

module.exports = function(grunt, options){

    const config = grunt.config.get('config');
    const glob = require('glob');

    // https://github.com/gruntjs/grunt-contrib-htmlmin
    return {

        options: {
            collapseWhitespace: true,
            conservativeCollapse: true,
            collapseBooleanAttributes: true,
            removeComments: true,
            removeCommentsFromCDATA: true,
            removeOptionalTags: true,
            minifyCSS: true,
            expand: true
        },
        main: {
            // 모든 template html 을 압축한다.
            files: [{
                expand: true,
                cwd: '<%=config.tools.path.src%>',
                src: '<%=config.tools.relativePath.app%>/**/*.tpl.html',
                dest: '<%=config.tools.path.build%>'
            }]
        },
        index: {
            // build 된 index.html 을 압축한다.
            files: [{
                expand: true,
                cwd: '<%=config.tools.path.build%>',
                src: 'index.html',
                dest: '<%=config.tools.path.build%>'
            }]
        }
    };
};
