// bx-builder-angular/tools/grunt/config/sass.js
/**
 * Created by sgjeon on 16. 4. 8..
 */

const util = require("../../lib/util");
const _ = require('lodash');
const path = require('path');

module.exports = function(grunt, options){

    const config = grunt.config.get('config');

    // https://github.com/sindresorhus/grunt-sass
    const sass = {

        dev: {
            options: {
                sourceMap: false,
                outputStyle: 'nested'
            },
            files: {}
        },
        prod: {
            options: {
                sourceMap: false,
                outputStyle: 'compressed'
            },
            files: {}
        }
    };

    // sass files 을 초기화 한다.
    (function initSassFiles(){

        if (!config.build['css']) return;

        var css = config.build['css'];
        var files = css['files'] || css['external'];

        if (files){

            _.forEach(files, v => {

                var fileInfo = util.getFileInfo(v);

                var fileName = fileInfo.name;
                var _path = fileInfo.path;

                _path = util.replaceCSSPathToBuildAssetsPath(_path);

                var destName = util.replaceSassToCss(fileName);

                sass.dev.files[path.join(_path, destName)] = v;
                sass.prod.files[path.join(_path, destName)] = v;
            });
        }

    })();

    return sass;
};
