// eslint-plugin-bxuipAngular/src/rules/module-name.js

const utils = require('../lib/utils');
const msg = require('../lib/msg.json');

module.exports = {
    meta: {
        id: 'bxuipAngular/module-name',
        schema: [{
            type: ['string', 'object']
        }]
    },
    create: function(context) {

        var ruleMsg = msg['ko'][this.meta.id]['default'];

        return {

            CallExpression: function(node) {
                var prefix = context.options[0] || '/^[a-z][a-zA-Z0-9]*$/';

                var convertedPrefix = utils.convertPrefixToRegex(prefix);

                if (!utils.isStringRegexp(prefix)) return;

                if (utils.isAngularModuleDeclaration(node)) {

                    var name = node.arguments[0].value;

                    if (name && !convertedPrefix.test(name)) {

                        context.report(node, ruleMsg, {
                            module: name,
                            prefix: prefix.toString()
                        });
                    }
                }
            }
        };
    }
};
